x/superfluid/keeper/unpool.go

```diff
@@ -0,0 +1,178 @@
+	// 6) Make 1 new lock for every asset in collateral. Many code paths need this assumption to hold
+	// 7) Make new lock begin unlocking
+
+	// 0) check if pool is whitelisted for unpool
+	err := k.checkUnpoolWhitelisted(ctx, poolId)
+	if err != nil {
+		return []uint64{}, err
+	}
+
+	// 1) Consistency check that lockID corresponds to sender, and contains correct LP shares.
+	// These are expected to be true by the caller, but good to double check
+	// TODO: Try to minimize dependence on lock here
+	lock, err := k.validateLockForUnpool(ctx, sender, poolId, lockId)
+	if err != nil {
+		return []uint64{}, err
+	}
+
+	// 2) Get remaining duration on the lock. Handle if the lock was unbonding.
+	lockRemainingDuration := k.getExistingLockRemainingDuration(ctx, lock)
+
+	// 3) If superfluid delegated, superfluid undelegate
+	err = k.unbondSuperfluidIfExists(ctx, sender, lockId)
+	if err != nil {
+		return []uint64{}, err
+	}
+
+	// finish unlocking directly for locked locks
+	// this also unlocks locks that were in the unlocking queue
+	err = k.lk.ForceUnlock(ctx, *lock)
+	if err != nil {
+		return []uint64{}, err
+	}
+
+	// 4) ExitPool with these unlocked LP shares
+	gammShares := lock.Coins[0]
+	minOutCoins := sdk.NewCoins()
+	exitedCoins, err := k.gk.ExitPool(ctx, sender, poolId, gammShares.Amount, minOutCoins)
+	if err != nil {
+		return []uint64{}, err
+	}
+
+	// Make one new lock for every coin exited from the pool.
+	newLocks := make([]lockuptypes.PeriodLock, 0, len(exitedCoins))
+	newLockIds := make([]uint64, 0, len(exitedCoins))
+	for _, exitedCoin := range exitedCoins {
+		newLock, err := k.lk.LockTokens(ctx, sender, sdk.NewCoins(exitedCoin), lockRemainingDuration)
+		if err != nil {
+			return []uint64{}, err
+		}
+		newLocks = append(newLocks, newLock)
+		newLockIds = append(newLockIds, newLock.ID)
+	}
+
+	// 7) Begin unlocking every new lock
+	for _, newLock := range newLocks {
+		err = k.lk.BeginForceUnlock(ctx, newLock.ID, newLock.Coins)
+		if err != nil {
+			return []uint64{}, err
+		}
+	}
+
+	return newLockIds, nil
+}
+
+// check if pool is whitelisted for unpool
+func (k Keeper) checkUnpoolWhitelisted(ctx sdk.Context, poolId uint64) error {
+	allowedPools := k.GetUnpoolAllowedPools(ctx)
+	allowed := false
+
+	for _, allowedPoolId := range allowedPools {
+		if poolId == allowedPoolId {
+			allowed = true
+			break
+		}
+	}
+
+	if !allowed {
+		return types.ErrPoolNotWhitelisted
+	}
+
+	return nil
+}
+
+// check if pool is whitelisted for unpool
+func (k Keeper) validateLockForUnpool(ctx sdk.Context, sender sdk.AccAddress, poolId uint64, lockId uint64) (*lockuptypes.PeriodLock, error) {
+	lock, err := k.lk.GetLockByID(ctx, lockId)
+	if err != nil {
+		return lock, err
+	}
+
+	// consistency check: validate lock owner
+	// However, we expect this to be guaranteed by caller though.
+	if lock.Owner != sender.String() {
+		return lock, lockuptypes.ErrNotLockOwner
+	}
+
+	gammShare := lock.Coins[0]
```

Comment:
Is the length of locks.Coins always 1?

Reply:
yeah, this is gauranteed by the validate function, I'll add a comment and link

Reply:
Yup `GetLockById(...)` only returns 1 lock in this format 
`ID:1 owner:"osmo1v9jxgu3395kj6tfd95kj6tfd95kj6tfd0hptum" duration:<seconds:1 > end_time:<seconds:-62135596800 > coins:<denom:"stake" amount:"10" > `